File: module_1/java/src/main/java/org/example/PullConsumerApp.java
package org.example;

import org.apache.kafka.clients.consumer.ConsumerConfig;
import org.apache.kafka.clients.consumer.ConsumerRecord;
import org.apache.kafka.clients.consumer.ConsumerRecords;
import org.apache.kafka.clients.consumer.KafkaConsumer;
import org.apache.kafka.clients.producer.ProducerConfig;
import org.apache.kafka.common.serialization.StringDeserializer;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.time.Duration;
import java.util.List;
import java.util.Properties;

/**
 * This consumer manually polls Kafka for new messages, implementing a pull model
 */
public class PullConsumerApp {
    private static final Logger logger = LoggerFactory.getLogger(PullConsumerApp.class);

    public static void main(String[] args) {
        Properties props = new Properties();
        props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:29092,localhost:39092,localhost:49092");
        props.put(ConsumerConfig.GROUP_ID_CONFIG, "consumer-group-1");
        props.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName());
        props.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName());
        props.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, "false"); //отключает автоматическое сохранение смещения
        props.put(ConsumerConfig.FETCH_MIN_BYTES_CONFIG, 10 * 1024 * 1024); // для большей наглядности ограничен минимальный размер пакета сообщений

        KafkaConsumer<String, String> consumer = new KafkaConsumer<>(props);
        consumer.subscribe(List.of("test-topic"));

        try {
            while (true) {
                ConsumerRecords<String, String> records = consumer.poll(Duration.ofMillis(10000));
                for (ConsumerRecord<String, String> record : records) {
                    Message message = Message.deserialize(record.value());
                    logger.info("Pulled Message: " + message.getContent());
                    consumer.commitSync();
                }
            }
        } catch (Exception e) {
            logger.error("Failure", e);
        } finally {
            consumer.close();
        }
    }
}

File: module_1/project/python/src/pull_consumer_app.py
import logging.config

from confluent_kafka import Consumer

from models.message import Message
from settings.logging import LOGGING_CONFIG

logging.config.dictConfig(LOGGING_CONFIG)
logger = logging.getLogger("pull_consumer_app")


def pull_consume():
    """
    This consumer manually polls Kafka for new messages, implementing a pull model.
    """
    consumer_conf = {
        "bootstrap.servers": "localhost:29092,localhost:39092,localhost:49092",
        "group.id": "consumer-group-1",
        "auto.offset.reset": "earliest",
        "enable.auto.commit": False,          # Отключает автоматическое сохранение смещения
        "fetch.min.bytes": 10 * 1024 * 1024,  # Для большей наглядности ограничен минимальный размер пакета сообщений
        "fetch.wait.max.ms": 10_000,
    }
    consumer = Consumer(consumer_conf)
    consumer.subscribe(["test-topic"])

    try:
        while True:
            records = consumer.consume(timeout=10)

            for record in records:
                message = Message.deserialize(record.value().decode("utf-8"))
                logger.info(f"Pulled Message: {message.content}")
            consumer.commit()
    except Exception as e:
        logger.exception(f"Failure: {e}")
    finally:
        consumer.close()


if __name__ == "__main__":
    pull_consume()
